server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateAttributesManager.java

```diff
@@ -0,0 +1,195 @@
+        String component,
+        AbstractRemoteWritableBlobEntity remoteObject,
+        LatchedActionListener<ClusterMetadataManifest.UploadedMetadata> latchedActionListener
+    ) {
+        return ActionListener.wrap(
+            resp -> latchedActionListener.onResponse(remoteObject.getUploadedMetadata()),
+            ex -> latchedActionListener.onFailure(new RemoteClusterStateUtils.RemoteStateTransferException(component, remoteObject, ex))
+        );
+    }
+
+    public CheckedRunnable<IOException> getAsyncMetadataReadAction(
+        String clusterUUID,
+        String component,
+        String componentName,
+        String uploadedFilename,
+        LatchedActionListener<RemoteReadResult> listener
+    ) {
+        final ActionListener actionListener = ActionListener.wrap(
+            response -> listener.onResponse(new RemoteReadResult((ToXContent) response, CLUSTER_STATE_ATTRIBUTE, component)),
+            listener::onFailure
+        );
+        if (component.equals(RemoteDiscoveryNodes.DISCOVERY_NODES)) {
+            RemoteDiscoveryNodes remoteDiscoveryNodes = new RemoteDiscoveryNodes(
+                uploadedFilename,
+                clusterUUID,
+                compressor,
+                namedXContentRegistry
+            );
+            return () -> discoveryNodesBlobStore.readAsync(remoteDiscoveryNodes, actionListener);
+        } else if (component.equals(RemoteClusterBlocks.CLUSTER_BLOCKS)) {
+            RemoteClusterBlocks remoteClusterBlocks = new RemoteClusterBlocks(
+                uploadedFilename,
+                clusterUUID,
+                compressor,
+                namedXContentRegistry
+            );
+            return () -> clusterBlocksBlobStore.readAsync(remoteClusterBlocks, actionListener);
+        } else if (component.equals(CLUSTER_STATE_CUSTOM)) {
+            final ActionListener customActionListener = ActionListener.wrap(
+                response -> listener.onResponse(
+                    new RemoteReadResult(
+                        (ToXContent) response,
+                        CLUSTER_STATE_ATTRIBUTE,
+                        String.join(CUSTOM_DELIMITER, component, componentName)
+                    )
+                ),
+                listener::onFailure
+            );
+            RemoteClusterStateCustoms remoteClusterStateCustoms = new RemoteClusterStateCustoms(
+                uploadedFilename,
+                componentName,
+                clusterUUID,
+                compressor,
+                namedXContentRegistry,
+                namedWriteableRegistry
+            );
+            return () -> customsBlobStore.readAsync(remoteClusterStateCustoms, customActionListener);
+        } else {
+            throw new RemoteStateTransferException("Remote object not found for " + component);
+        }
+    }
+
+    public Map<String, ClusterState.Custom> getUpdatedCustoms(ClusterState clusterState, ClusterState previousClusterState) {
```

Comment:
I hope this logic is at one place now and would be cleaned from RemoteClusterStateService